ext/websocket/lib.rs

```diff
@@ -426,6 +471,54 @@ pub async fn op_ws_send(
   Ok(())
 }
 
+#[op]
+pub async fn op_ws_send_string(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  text: String,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<WsStreamResource>(rid)?;
+  resource.send(Message::Text(text)).await?;
+  Ok(())
+}
+
+#[op]
+pub async fn op_ws_send_binary(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  data: ZeroCopyBuf,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<WsStreamResource>(rid)?;
+  resource.send(Message::Binary(data.to_vec())).await?;
+  Ok(())
+}
+
+#[op]
+pub fn op_ws_try_send_string(
+  state: &mut OpState,
+  rid: ResourceId,
+  text: String,
+) -> Result<bool, AnyError> {
+  let resource = state.resource_table.get::<WsStreamResource>(rid)?;
+  resource.try_send(Message::Text(text))
+}
+
+#[op(fast)]
+pub fn op_ws_try_send_binary(
+  state: &mut OpState,
+  rid: u32,
+  value: &[u8],
+) -> Result<bool, AnyError> {
+  let resource = state.resource_table.get::<WsStreamResource>(rid)?;
+  resource.try_send(Message::Binary(value.to_vec()))
```

Comment:
Sad that it's not possible to just use the slice for this synchronous call :/

**Nit:** There's a `From<&'b [u8]>` implementation for `Message`.

Reply:
Also a `From<String>` apparently.

Reply:
oh its indeed possible to assume `'static` lifetime for that slice since this is a sync call.

Reply:
Oh yeah. I couldn't find a method for it directly but I guess it should be possible to just do:
```rs
let vec = Vec::from_raw_parts::<u8>(value.as_ptr(), value.len(), value.len());
```
and avoid the copy.

Reply:
umm we cant do that. tungstenite will free the vec owned by v8

blocked on https://github.com/snapview/tungstenite-rs/pull/175

Reply:
Oh very true. Note to self: Don't lie to Rust about ownership of structs.